test/cctest/test_environment.cc

```diff
@@ -241,3 +306,144 @@ TEST_F(EnvironmentTest, BufferWithFreeCallbackIsDetached) {
   CHECK_EQ(callback_calls, 1);
   CHECK_EQ(ab->ByteLength(), 0);
 }
+
+#if HAVE_INSPECTOR
+TEST_F(EnvironmentTest, InspectorMultipleEmbeddedEnvironments) {
+  // Tests that child Environments can be created through the public API
+  // that are accessible by the inspector.
+  // This test sets a global variable in the child Environment, and reads it
+  // back both through the inspector and inside the child Environment, and
+  // makes sure that those correspond to the value that was originally set.
+  const v8::HandleScope handle_scope(isolate_);
+  const Argv argv;
+  Env env {handle_scope, argv};
+
+  v8::Local<v8::Context> context = isolate_->GetCurrentContext();
+  node::LoadEnvironment(*env,
+      "'use strict';\n"
+      "const { Worker } = require('worker_threads');\n"
+      "const { Session } = require('inspector');\n"
+
+      "const session = new Session();\n"
+      "session.connect();\n"
+      "session.on('NodeWorker.attachedToWorker', (\n"
+      "  ({ params: { workerInfo, sessionId } }) => {\n"
+      "    session.post('NodeWorker.sendMessageToWorker', {\n"
+      "      sessionId,\n"
+      "      message: JSON.stringify({\n"
+      "        id: 1,\n"
+      "        method: 'Runtime.evaluate',\n"
+      "        params: {\n"
+      "          expression: 'global.variableFromParent = 42;'\n"
+      "        }\n"
+      "      })\n"
+      "    });\n"
+      "    session.on('NodeWorker.receivedMessageFromWorker',\n"
+      "      ({ params: { message } }) => {\n"
+      "        global.messageFromWorker = \n"
+      "          JSON.parse(message).result.result.value;\n"
+      "      });\n"
+      "  }));\n"
+      "session.post('NodeWorker.enable', { waitForDebuggerOnStart: false });\n")
+          .ToLocalChecked();
+
+  struct ChildEnvironmentData {
+    node::ThreadId thread_id;
+    std::unique_ptr<node::InspectorParentHandle> inspector_parent_handle;
+    node::MultiIsolatePlatform* platform;
+    int32_t extracted_value = -1;
+    uv_async_t thread_stopped_async;
+  };
+
+  ChildEnvironmentData data;
+  data.thread_id = node::AllocateEnvironmentThreadId();
+  data.inspector_parent_handle =
+      GetInspectorParentHandle(*env, data.thread_id, "file:///embedded.js");
```

Comment:
> From the look of the test I still cannot see where the necessity that the user must call `node::AllocateEnvironmentThreadId()` and `GetInspectorParentHandle ()` here is from (other than "this is currently how the lifetime is managed internally") - if the URL and the thread ID are essentially tied to a Node.js instance, why do they need to be managed separately and explicitly, independent of the Environment methods?

I’m not sure I really understand. There basically has to be a call on the “parent” Environment’s thread to get data that can then be used on the “child” Environment’s thread, if they run on different threads, otherwise you’d run into race conditions – no?

> What would happen if `node::CreateEnvironment` is called with a thread id that is different from the one assigned to the the parent handle later passed to `node::LoadEnvironmet`? 

I’m guessing that the inspector would not work as expected, or maybe just report a wrong id? I don’t know for sure.

> Can we just make these part of `Environment` and hide them from users?

I’m still not sure what you have in mind for this.

Reply:
> There basically has to be a call on the “parent” Environment’s thread to get data that can then be used on the “child” Environment’s thread

I think I am essentially asking what kind of data that is allocated here and if it's inevitable that it has to be allocated on the parent thread - for example could it be allocated with a reference to the parent Environment on the child thread, so that the allocation can be hidden as an implementation detail? About the race, AFAICT, the current API requires user to allocate the parent handle on the main thread, and then use it on the child thread, is it really less racy compared to allocating and using the parent in one go in one thread under the API so we can apply locks ourselves and keep the read and write as close as possible?

Reply:
@joyeecheung I guess we could introduce locks here, but I don’t really see the need to do that/how that would simplify or otherwise improve the API?